### ISSUE TITLE
tr/// against $1 causes persistent data

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1964]
Migrated from [rt.perl.org#3237](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3237) (status was 'resolved')

Searchable as RT3237$

**Comment by p5pRT** [comment]
### From rootbeer@redcat.com

This is a bug report for perl from rootbeer@&#8203;redcat\.com\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.



---
The substitution near the end of this chunk of code should change $\_ from
'fred' to 'Fred'\, but a previous value from $1 leaks in as well\. This
seems to happen only when $1 has been used in a tr///\-for\-counting
operation\.

&nbsp;    $\_ = "fred";
&nbsp;    /\(\[a\-z\]\{2\}\)/;
&nbsp;    $1 =~ tr/A\-Z//;
&nbsp;    print;            \# prints "fred"
&nbsp;    s/^\(\\s\*\)f/$1F/;
&nbsp;    print;            \# prints "frFred"

It's entertaining to run this to completion in the debugger \(not stepping
through it\, but just letting it run\) then to read the "Debugged program
terminated" message\. Okay\, so it doesn't take much to amuse me\.  :\-D



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl v5\.6\.0&#8203;:

Configured by rootbeer at Sun Apr 23 14&#8203;:09&#8203;:13 PDT 2000\.

Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.0\.35\, archname=i586\-linux
&nbsp;    uname='linux localhost\.localdomain 2\.0\.35 \#1 tue jul 14 23&#8203;:56&#8203;:39 edt 1998 i586 unknown '
&nbsp;    config\_args=''
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-g \-DDEBUGGING \-O2'\, gccversion=2\.7\.2\.3
&nbsp;    cppflags=''
&nbsp;    ccflags =' \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
&nbsp;    libc=/lib/libc\-2\.0\.7\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.6\.0&#8203;:
&nbsp;    /usr/lib/perl5/5\.6\.0/i586\-linux
&nbsp;    /usr/lib/perl5/5\.6\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.6\.0/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl v5\.6\.0&#8203;:
&nbsp;    HOME=/home/rootbeer
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/usr/games&#8203;:/home/rootbeer/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/local/bin/bash

**Comment by p5pRT** [comment]
### From @vanstyn

In \<Pine\.GSO\.4\.10\.10005111759110\.16364\-100000@&#8203;user2\.teleport\.com>\, Tom Phoenix
writes&#8203;:
:This is a bug report for perl from rootbeer@&#8203;redcat\.com\,
:generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
:
:\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
:
:The substitution near the end of this chunk of code should change $\_ from
:'fred' to 'Fred'\, but a previous value from $1 leaks in as well\. This
:seems to happen only when $1 has been used in a tr///\-for\-counting
:operation\.
:
:    $\_ = "fred";
:    /\(\[a\-z\]\{2\}\)/;
:    $1 =~ tr/A\-Z//;
:    print;            \# prints "fred"
:    s/^\(\\s\*\)f/$1F/;
:    print;            \# prints "frFred"

This appears to occur because the tr/// sets POK on $1\, which bypasses
the magic get that should happen in the substitution\. The culprit is
this line from doop\.c&#8203;:595 in Perl\_do\_trans\(\)&#8203;:
&nbsp;    \(void\)SvPOK\_only\(sv\);

I'm not sure quite why this line exists\, since removing it does not
trigger any test failures\, but I suspect that it should occur only if
the SV does not have a magic get \(as in the attached patch\)\.
If someone can confirm or correct this belief\, I'll aim to add
appropriate test cases\.

Hugo
<details><summary>Inline Patch</summary>

```diff
--- doop.c.old	Sun Mar 12 03:36:32 2000
+++ doop.c	Fri May 12 10:41:04 2000
@@ -592,7 +592,8 @@
 	return 0;
     if (!SvPOKp(sv))
 	(void)SvPV_force(sv, len);
-    (void)SvPOK_only(sv);
+    if (!SvGMAGICAL(sv))
+	(void)SvPOK_only(sv);
 
     DEBUG_t( Perl_deb(aTHX_ "2.TBL\n"));
 
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Hugo \<hv@&#8203;crypt\.compulink\.co\.uk> wrote
> This appears to occur because the tr/// sets POK on $1\, which bypasses
> the magic get that should happen in the substitution\. The culprit is
> this line from doop\.c&#8203;:595 in Perl\_do\_trans\(\)&#8203;:
>     \(void\)SvPOK\_only\(sv\);
> 
> I'm not sure quite why this line exists\, since removing it does not
> trigger any test failures\, but I suspect that it should occur only if
> the SV does not have a magic get \(as in the attached patch\)\.

It exists to ensure that the attached tests pass\.   :\-\)

And I think the "correct" fix is to avoid the SvPOK\_only in the
non\-updating case\.    Compare with the read\-only test just above\.

Patch for 5\.6\.0 and new tests attached\.

And I've added comments to the new tests  \-  perhaps that'll start
a trend\.       :\-\)


Mike Guy

<details><summary>Inline Patch</summary>

```diff
--- ./t/op/tr.t.orig	Fri May 12 17:28:34 2000
+++ ./t/op/tr.t	Fri May 12 17:34:18 2000
@@ -5,7 +5,7 @@
     unshift @INC, "../lib";
 }
 
-print "1..4\n";
+print "1..6\n";
 
 $_ = "abcdefghijklmnopqrstuvwxyz";
 
@@ -37,3 +37,19 @@
     print "ok 4\n";
 }
 #
+
+# make sure that tr cancels IOK and NOK
+($x = 12) =~ tr/1/3/;
+(my $y = 12) =~ tr/1/3/;
+($f = 1.5) =~ tr/1/3/;
+(my $g = 1.5) =~ tr/1/3/;
+print "not " unless $x + $y + $f + $g == 71;
+print "ok 5\n";
+
+# make sure tr is harmless if not updating  -  see [ID 20000511.005]
+$_ = 'fred';
+/([a-z]{2})/;
+$1 =~ tr/A-Z//;
+s/^(\s*)f/$1F/;
+print "not " if $_ ne 'Fred';
+print "ok 6\n";
--- ./doop.c.orig	Fri May 12 15:41:40 2000
+++ ./doop.c	Fri May 12 17:40:21 2000
@@ -592,7 +592,8 @@
 	return 0;
     if (!SvPOKp(sv))
 	(void)SvPV_force(sv, len);
-    (void)SvPOK_only(sv);
+    if (!(PL_op->op_private & OPpTRANS_IDENTICAL))
+	(void)SvPOK_only(sv);
 
     DEBUG_t( Perl_deb(aTHX_ "2.TBL\n"));
 

End of patch
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I wrote
> And I think the "correct" fix is to avoid the SvPOK\_only in the
> non\-updating case\.

For some values of "correct" only\.    This second try ensures that
the UTF8 flag doesn't get lost\.    \(But tr/// on UTF8 values is broken
in other ways\.    And the new tests are somewhat curiously written to
circumvent UTF8 bugs in eq and other operators\.   But that's all the
subject of another thread\.\)

I note that Robin Barker\, in his recent patch to vec\(\)\, uses SvNIOK\_off
rather than SvPOK\_only or SvPOK\_only\_UTF8\.     I'm unclear which is
more correct\.    I can't find any case which fails with my test \[\*\]; in
partticular tied values are treated correctly although I don't
understand why the magic isn't getting lost\.

\[\*\]  Though I did find some strange examples of "correct" behaviour\.
&nbsp;     Can you predict what

&nbsp;            \($\! = 28\) =~ tr/2/3/

&nbsp;     will do?


Mike Guy

<details><summary>Inline Patch</summary>

```diff
--- ./t/op/tr.t.orig	Fri May 12 19:06:02 2000
+++ ./t/op/tr.t	Sat May 13 09:48:39 2000
@@ -5,7 +5,7 @@
     unshift @INC, "../lib";
 }
 
-print "1..4\n";
+print "1..8\n";
 
 $_ = "abcdefghijklmnopqrstuvwxyz";
 
@@ -37,3 +37,27 @@
     print "ok 4\n";
 }
 #
+
+# make sure that tr cancels IOK and NOK
+($x = 12) =~ tr/1/3/;
+(my $y = 12) =~ tr/1/3/;
+($f = 1.5) =~ tr/1/3/;
+(my $g = 1.5) =~ tr/1/3/;
+print "not " unless $x + $y + $f + $g == 71;
+print "ok 5\n";
+
+# make sure tr is harmless if not updating  -  see [ID 20000511.005]
+$_ = 'fred';
+/([a-z]{2})/;
+$1 =~ tr/A-Z//;
+s/^(\s*)f/$1F/;
+print "not " if $_ ne 'Fred';
+print "ok 6\n";
+
+# check tr handles UTF8 correctly
+($x = 256.65.258) =~ tr/a/b/;
+print "not " if $x ne 256.65.258 or length $x != 3;
+print "ok 7\n";
+$x =~ tr/A/B/;
+print "not " if $x ne 256.66.258 or length $x != 3;
+print "ok 8\n";
--- ./doop.c.orig	Fri May 12 19:06:03 2000
+++ ./doop.c	Sat May 13 09:54:44 2000
@@ -592,7 +592,8 @@
 	return 0;
     if (!SvPOKp(sv))
 	(void)SvPV_force(sv, len);
-    (void)SvPOK_only(sv);
+    if (!(PL_op->op_private & OPpTRANS_IDENTICAL))
+	(void)SvPOK_only_UTF8(sv);
 
     DEBUG_t( Perl_deb(aTHX_ "2.TBL\n"));
 

End of patch
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

M\.J\.T\. Guy writes&#8203;:
>      Can you predict what
>             \($\! = 28\) =~ tr/2/3/
>      will do?

Lemme try to predict what it \*should\* do\.  $\! is assignable as a
number only\, but accessible any way\.  So tr/// is performed on the
message of errno==28\.  But then the converted message is either
assigned back\, or \(as an optimization\) the assignment is skipped
\(since most probably there is no '2' in the message\)\.

In the first case numerical value of the message \(thus 0\) is put in
$\!\.  In the second case tr/// is a NOP\.

Checking\.\.\.  I get

&nbsp;  Argument "Disk full" isn't numeric in trans at \-e line 1\.

So there is no skip\-assignment optimization\.\.\.

Ilya

**Comment by p5pRT** [comment]
### From @smpeters

> \[RT\_System \- Thu May 11 20&#8203;:40&#8203;:52 2000\]&#8203;:
> 
> In \<Pine\.GSO\.4\.10\.10005111759110\.16364\-100000@&#8203;user2\.teleport\.com>\, Tom
> Phoenix
> writes&#8203;:
> :This is a bug report for perl from rootbeer@&#8203;redcat\.com\,
> :generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> :
> :\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> :
> :The substitution near the end of this chunk of code should change $\_
> from
> :'fred' to 'Fred'\, but a previous value from $1 leaks in as well\. This
> :seems to happen only when $1 has been used in a tr///\-for\-counting
> :operation\.
> :
> :    $\_ = "fred";
> :    /\(\[a\-z\]\{2\}\)/;
> :    $1 =~ tr/A\-Z//;
> :    print;            \# prints "fred"
> :    s/^\(\\s\*\)f/$1F/;
> :    print;            \# prints "frFred"
> 
> This appears to occur because the tr/// sets POK on $1\, which bypasses
> the magic get that should happen in the substitution\. The culprit is
> this line from doop\.c&#8203;:595 in Perl\_do\_trans\(\)&#8203;:
>     \(void\)SvPOK\_only\(sv\);
> 
> I'm not sure quite why this line exists\, since removing it does not
> trigger any test failures\, but I suspect that it should occur only if
> the SV does not have a magic get \(as in the attached patch\)\.
> If someone can confirm or correct this belief\, I'll aim to add
> appropriate test cases\.
> 
> Hugo
> \-\-\- doop\.c\.old    Sun Mar 12 03&#8203;:36&#8203;:32 2000
> \+\+\+ doop\.c    Fri May 12 10&#8203;:41&#8203;:04 2000
> @&#8203;@&#8203; \-592\,7 \+592\,8 @&#8203;@&#8203;
>      return 0;
>      if \(\!SvPOKp\(sv\)\)
>      \(void\)SvPV\_force\(sv\, len\);
> \-    \(void\)SvPOK\_only\(sv\);
> \+    if \(\!SvGMAGICAL\(sv\)\)
> \+    \(void\)SvPOK\_only\(sv\);
> 
>      DEBUG\_t\( Perl\_deb\(aTHX\_ "2\.TBL\\n"\)\);
> 

It looks like this patch or the other patch in this ticket thread were
never applied\.  Instead\, the following change seemed to take care of things\.

Change 17984 by rgs@&#8203;rgs\-home on 2002/10/09 19&#8203;:17&#8203;:08

&nbsp;    Fix bug \#17823 : non\-modifying tr/// stringifies references

Affected files \.\.\.

\.\.\. //depot/perl/doop\.c\#129 edit
\.\.\. //depot/perl/t/op/tr\.t\#33 edit

Differences \.\.\.

==== //depot/perl/doop\.c\#129 \(text\) ====

@&#8203;@&#8203; \-608\,10 \+608\,11 @&#8203;@&#8203;
&nbsp;     \(void\)SvPV\(sv\, len\);
&nbsp;     if \(\!len\)
&nbsp;     return 0;
\-    if \(\!SvPOKp\(sv\)\)
\-    \(void\)SvPV\_force\(sv\, len\);
\-    if \(\!\(PL\_op\->op\_private & OPpTRANS\_IDENTICAL\)\)
\+    if \(\!\(PL\_op\->op\_private & OPpTRANS\_IDENTICAL\)\) \{
\+    if \(\!SvPOKp\(sv\)\)
\+        \(void\)SvPV\_force\(sv\, len\);
&nbsp;     \(void\)SvPOK\_only\_UTF8\(sv\);
\+    \}
 
&nbsp;     DEBUG\_t\( Perl\_deb\(aTHX\_ "2\.TBL\\n"\)\);
 

==== //depot/perl/t/op/tr\.t\#33 \(xtext\) ====

@&#8203;@&#8203; \-6\,7 \+6\,7 @&#8203;@&#8203;
&nbsp;     require '\./test\.pl';
 \}
 
\-plan tests => 97;
\+plan tests => 99;
 
 my $Is\_EBCDIC = \(ord\('i'\) == 0x89 & ord\('J'\) == 0xd1\);
 
@&#8203;@&#8203; \-379\,3 \+379\,7 @&#8203;@&#8203;
 eval '$foo\{bar\} =~ tr/N/N/';
 is\( $@&#8203;\, ''\,         'implicit count outside hash bounds' \);
 is\( scalar keys %foo\, 0\,   "    doesn't extend the hash"\);
\+
\+$x = \\"foo";
\+is\( $x =~ tr/A/A/\, 2\, 'non\-modifying tr/// on a scalar ref' \);
\+is\( ref $x\, 'SCALAR'\, "    doesn't stringify its argument" \);

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
